tests/unit/test_layers.py

```diff
@@ -0,0 +1,332 @@
+    expected_tar_contents = [
+        "first",
+        "first/first.txt",
+        "second",
+        "second/.wh..wh..opq",
+        "second/second.txt",
+    ]
+    assert sorted(temp_tar_contents) == expected_tar_contents
+
+
+def test_archive_layer_with_base_layer_subdirs(tmp_path):
+    """Test base layer handling with subdirectories."""
+    layer_dir = tmp_path / "layer_dir"
+    layer_dir.mkdir()
+
+    # The path "second/" contains multiple subdirectories.
+    (layer_dir / "second").mkdir()
+    (layer_dir / "second/second.txt").touch()
+
+    (layer_dir / "second/subdir").mkdir()
+    (layer_dir / "second/subdir/subdir_file.txt").touch()
+
+    (layer_dir / "second/subdir/subsubdir").mkdir()
+    (layer_dir / "second/subdir/subsubdir/subsubdir_file.txt").touch()
+
+    (layer_dir / "third").mkdir()
+    (layer_dir / "third/third.txt").touch()
+
+    rootfs_dir = tmp_path / "rootfs"
+    rootfs_dir.mkdir()
+
+    # The base layer has a "first/" dir and a symlink to it called "second".
+    # Every subdirectory in "second/" in the "upper" layer must be added as
+    # a subdir of "first".
+    (rootfs_dir / "first").mkdir()
+    (rootfs_dir / "second").symlink_to("first")
+
+    temp_tar_path = tmp_path / "layer.tar"
+    layers.archive_layer(layer_dir, temp_tar_path, base_layer_dir=rootfs_dir)
+    temp_tar_contents = get_tar_contents(temp_tar_path)
+
+    expected_tar_contents = [
+        "first/second.txt",
+        "first/subdir",
+        "first/subdir/subdir_file.txt",
+        "first/subdir/subsubdir",
+        "first/subdir/subsubdir/subsubdir_file.txt",
+        "third",
+        "third/third.txt",
+    ]
+    assert temp_tar_contents == expected_tar_contents
+
+
+def test_archive_layer_duplicate_dirs(tmp_path):
+    """
+    Test creating a layer where, because of symlinks in the base, multiple
+    directories end up as the same target.
+    """
+
+    layer_dir, rootfs_dir = duplicate_dirs_setup(tmp_path)
+
+    temp_tar_path = tmp_path / "layer.tar"
+    layers.archive_layer(layer_dir, temp_tar_path, base_layer_dir=rootfs_dir)
+    temp_tar_contents = get_tar_contents(temp_tar_path)
+
+    expected_tar_contents = [
+        "usr",
+        "usr/bin",
+        "usr/bin/dir1",
+        "usr/bin/dir1/a.txt",
+        "usr/bin/dir1/b.txt",
+    ]
+    assert temp_tar_contents == expected_tar_contents
+
+
+def test_archive_layer_duplicate_dirs_conflict(tmp_path):
+    """
+    Test creating a layer where, because of symlinks in the base, multiple
+    directories end up as the same target but the directories have different
+    ownership/permissions.
+    """
+    layer_dir, rootfs_dir = duplicate_dirs_setup(tmp_path)
+
+    # Change the default permissions of the directories that will end up as
+    # "/usr/bin/dir1", to ensure that they are different.
+    if sys.platform == "win32":
+        # On Windows, chmod() can only change a directory's read-only status.
+        (layer_dir / "bin/dir1").chmod(stat.S_IWRITE)
```

Comment:
ok I lied; when porting the tests over I also took the opportunity to ensure that these tests also work on Windows (they were skipped before because the whole test_oci.py module skips Windows). This is a change to support that.

Reply:
nice